### PR TITLE
Optimisation demo: checks and plotting

### PR DESCRIPTION
**Pull request by jwallwork23**
Merges into #320.

This PR adds a few nice things for the optimisation example:
* Check both the Python and Fortran implementations converge to the expected array [1,2,3,4].
* Track the convergence of the loss function for both implementations and write these to files.
* Add a `ctest` check that the plotting script can be run and produces a PNG file.

### CODE REVIEW

**Comment by jwallwork23** [comment]
losses

For reference, this is what `losses.png` looks like for me. PyTorch and FTorch have exactly the same behaviour, as we might hope!

**Comment by jwallwork23** [comment]
For reference, this is the plot I get. PyTorch and FTorch have exactly the same behaviour, as we might hope!

losses

**Comment by jwallwork23** [comment]
> Hey @jwallwork23 could you rebase on `optim` to make this cleaner and pick up latest changes, then we can hopefully merge.
> 
> I'm happy with the png if you are - though suggest making the label-size of the `X` larger, or plotting them at the top level, so it is clear that both are on there.

Thanks @jatkinson1000. Reordered the plot symbols so that the cross is on top of the circle. The rebase picked up some slight differences and I propose some minor rewordings of the NOTE comments, which you might want to check you're happy with.

File: .gitignore [hunk 1]
```diff
@@ -196,4 +196,7 @@ cython_debug/
 *.pt
 
 # Apple Desktop Services Store
-**.DS_Store
\ No newline at end of file
+**.DS_Store
+
+# Example outputs
+examples/n_Optimizers/losses.png
```
Comment (jatkinson1000) [inline .gitignore]: Is it worth just ignoring `*.png` files and explicitly adding when we need given this path will be changing soon?

File: examples/n_Optimizers/expected_losses.png
Comment (jatkinson1000) [inline examples/n_Optimizers/expected_losses.png:1]: Your plotting changes don't seem to be reflected in this file (unless it's a caching problem...)

File: examples/n_Optimizers/optimizers.f90 [hunk 2]
```diff
@@ -45,32 +48,40 @@ program foptimizer
...
+  scaling_data(:) = 1.0_wp
+  call torch_tensor_from_array(scaling_tensor, scaling_data, tensor_layout, torch_kCPU, &
+                               requires_grad=.true.)
 
   ! Initialise an optimizer and apply it to scaling_tensor
   ! NOTE: The optimizer expects an array of tensors.
   call torch_optim_SGD(optimizer, [scaling_tensor], learning_rate=1D0)
 
+  ! Create an empty tensor for the gradient of the scaling
+  call torch_tensor_empty(scaling_grad, ndims, tensor_shape, torch_kFloat32, torch_kCPU)
```
Comment (jatkinson1000) [inline examples/n_Optimizers/optimizers.f90]: Would suggest leaving this next to the creation of scaling tensor (lines 55-56) instead of moving here, unless you had a reason?

File: examples/n_Optimizers/optimizers.f90 [hunk 3]
```diff
@@ -45,32 +48,40 @@ program foptimizer
...
   call torch_tensor_from_array(target_vec, target_data, tensor_layout, torch_kCPU)
 
-  ! Initialise Scaling tensor as ones as in Python example and a tensor for its gradient
-  call torch_tensor_ones(scaling_tensor, ndims, tensor_shape, &
-                         torch_kFloat32, torch_kCPU, requires_grad=.true.)
-  call torch_tensor_empty(scaling_grad, ndims, tensor_shape, torch_kFloat32, torch_kCPU)
+  ! Initialise Scaling tensor as ones as in Python example
+  scaling_data(:) = 1.0_wp
+  call torch_tensor_from_array(scaling_tensor, scaling_data, tensor_layout, torch_kCPU, &
+                               requires_grad=.true.)
```
Comment (jatkinson1000) [inline examples/n_Optimizers/optimizers.f90:54]: Thanks for spotting this needs to be a layout!

Comment (jwallwork23) [reply examples/n_Optimizers/optimizers.f90:54]: Ah sorry, must've picked the wrong change in the rebase. Fixed in 5ab663aca6ddd3e01f3fdadd70e8810a036a29a5.

File: examples/n_Optimizers/optimizers.f90 (hunk 2 above)
Comment (joewallwork) [reply examples/n_Optimizers/optimizers.f90]: Ah sorry, must've picked the wrong change in the rebase. Fixed in 5ab663aca6ddd3e01f3fdadd70e8810a036a29a5.

File: .gitignore (hunk 1 above)
Comment (joewallwork) [reply .gitignore]: Done in 5ab663aca6ddd3e01f3fdadd70e8810a036a29a5.

File: examples/n_Optimizers/expected_losses.png
Comment (joewallwork) [reply examples/n_Optimizers/expected_losses.png:1]: My bad - forgot to update it. Done in 5ab663aca6ddd3e01f3fdadd70e8810a036a29a5.
